web/src/main/java/it/pagopa/selfcare/dashboard/web/security/ExchangeTokenService.java

```diff
@@ -107,12 +111,123 @@ public ExchangedToken exchange(String institutionId, String productId, Optional<
-                    : ""
+                        ? workContract.getEmail().getValue()
+                        : ""
                 )
                 .orElse(null);
         claims.setEmail(email);
 
+        Institution institution = retrieveInstitution(institutionId, institutionInfo, List.of(productGrantedAuthority));
+        retrieveAndSetGroups(institution, institutionId, productId, principal);
+
+        claims.setInstitution(institution);
+        Product product = productsConnector.getProduct(productId);
+        environment.ifPresentOrElse(env -> claims.setAudience(product.getBackOfficeEnvironmentConfigurations().get(env).getIdentityTokenAudience())
+                , () -> claims.setAudience(product.getIdentityTokenAudience()));
+        claims.setDesiredExpiration(claims.getExpiration());
+        claims.setIssuedAt(new Date());
+        claims.setExpiration(Date.from(claims.getIssuedAt().toInstant().plus(duration)));
+        log.debug(LogUtils.CONFIDENTIAL_MARKER, "Exchanged claims = {}", claims);
+        String jwts = createJwts(claims);
+        log.debug(LogUtils.CONFIDENTIAL_MARKER, "Exchanged token = {}", jwts);
+        final String urlBO = environment.map(env -> product.getBackOfficeEnvironmentConfigurations().get(env).getUrl())
+                .orElse(product.getUrlBO());
+        log.trace("exchange end");
+        return new ExchangedToken(jwts, urlBO);
+    }
+
+    public ExchangedToken retrieveBillingExchangedToken(String institutionId) {
+
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        if (authentication == null) {
+            throw new IllegalStateException("Authentication is required");
+        }
+
+        List<Product> tempList = new ArrayList<>();
+        productsConnector.getProductsTree()
+                .forEach(productTree -> {
+                    tempList.add(productTree.getNode());
+                    if (!CollectionUtils.isEmpty(productTree.getChildren())) {
+                        tempList.addAll(productTree.getChildren());
+                    }
+                });
+
+        List<String> productList = tempList.stream()
+                .filter(Product::isInvoiceable)
+                .map(Product::getId)
+                .toList();
+
+        final List<ProductGrantedAuthority> productGrantedAuthorities = new ArrayList<>();
+
+        authentication.getAuthorities()
+                .stream()
+                .filter(grantedAuthority -> SelfCareGrantedAuthority.class.isAssignableFrom(grantedAuthority.getClass()))
+                .map(SelfCareGrantedAuthority.class::cast)
+                .filter(grantedAuthority -> institutionId.equals(grantedAuthority.getInstitutionId()))
+                .map(SelfCareGrantedAuthority::getRoleOnProducts)
+                .forEach(stringProductGrantedAuthorityMap -> checkIfIsInvoiceable(stringProductGrantedAuthorityMap, productList, productGrantedAuthorities));
```

Comment:
it is not clear what the method does....you could change method's name with such as "addProductIfIsInvoiceable"